Speedup environment activation

alalazo opened the pull request:
fixes #13554

alalazo commented:
@adamjstewart Can you try this and report back? I'll try to see if I can improve it further in terms of performance.

alalazo commented:
I think a refactor of `_set_variables_for_single_module` might also be a good thing, but given the entity of the performance bug reported in #13554 I'd be in  favor of merging this as a hotfix when considered done and do the refactor without hurry in a later PR.

adamjstewart commented:
Here are the same benchmarks on my laptop. From my home directory:
```console
$ time spack help
...
real	0m2.366s
user	0m0.728s
sys	0m0.298s
```
From the directory containing the large `spack.yaml` (installed environment):
```console
$ time spack help
...
real	0m6.649s
user	0m4.290s
sys	0m1.054s
```
Definitely a drastic improvement over `develop`!

alalazo commented:
@adamjstewart So far I tried to optimize for your small environment. Making a small break while the big one is building :slightly_smiling_face:

alalazo commented:
@adamjstewart Tried on the big environment. I might shave something more if we are fine losing the information in:

https://github.com/spack/spack/blob/1114ae93752aa9d33c3520f72aa701d8105c09dd/lib/spack/spack/util/environment.py#L338-L348

in cases where we modify the environment a lot. If need be I can compute the information only in debug mode, but if the timings are good enough without this change I would keep the PR as it is. Let me know.

adamjstewart commented:
I'm not sure how that function is used, so I don't know if we can skip it or not.

alalazo commented:
@adamjstewart If there are warnings on e.g. the same variable set two times in a DAG (so the first set won't have any effect) it will print context to tell you where both set took place. It would be good to keep it as it is, so if the performance is acceptable right now I'd avoid making any change.

adamjstewart commented:
Yeah, I guess it depends on how much of a difference it makes. The current PR drops us from 2 minutes to 6 seconds. If it would drop us from 6 seconds to 3 seconds, it would be worth it in my mind.

alalazo commented:
> The current PR drops us from 2 minutes to 6 seconds.

You didn't try the latest 2 commits :wink:

adamjstewart commented:
Okay, with the latest commit, I'm down to 4 seconds.

tgamblin commented on lib/spack/spack/build_environment.py:
```diff
@@ -489,19 +512,37 @@ def static_to_shared_library(static_lib, shared_lib=None, **kwargs):
     modules = parent_class_modules(pkg.__class__)
+    if context not in ('run', 'build'):
+        msg = 'context must be either "run" or "build"'
+        raise RuntimeError(msg)
```
This should really be a `ValueError`.

alalazo commented:
@tgamblin I reverted c2f6113 and implemented request number 1. Without that commit we retain most of the speedup so it makes sense also to me to look for further improvements later.

tgamblin commented:
@adamjstewart: how fast is this for you now?

adamjstewart commented:
On my laptop:

### Before (develop)

From my home directory:
```console
$ time spack help
...
real	0m2.871s
user	0m0.719s
sys	0m0.307s
```
From a directory containing a large `spack.yaml`:
```console
$ time spack help
...
real	3m22.219s
user	3m10.927s
sys	0m3.275s
```

### After (fixes/speedup_environment_activation)

From my home directory:
```console
$ time spack help
...
real	0m1.972s
user	0m0.682s
sys	0m0.271s
```
From a directory containing a large `spack.yaml`:
```console
$ time spack help
...
real	0m5.015s
user	0m3.607s
sys	0m0.612s
```

Let me know if you want benchmarks on Blue Waters where the problem is even more pronounced.
